### PR TITLE
update JS interop - remove JS typedef references

### PR DESCRIPTION
**Pull request by devoncarew**
- update JS interop - remove JS typedef references
- fix https://github.com/dart-lang/setup-dart/issues/95

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by devoncarew** [comment]
Compilation issues I'm seeing locally:

```
> dart compile js --enable-experiment=inline-class -olib/main.js lib/main.dart && dart tool/sig.dart --generate

lib/node/actions/core.dart:11:14:
Error: Inline class 'Core' is marked with a '@JS' annotation, but its representation type is not a valid JS interop type: 'InvalidType(<invalid>)'.
inline class Core {
             ^
lib/node/actions/core.dart:12:19:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external String getInput(String name);
                  ^
lib/node/actions/core.dart:14:17:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external void setOutput(String name, String value);
                ^
lib/node/actions/core.dart:16:17:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external void info(String name);
                ^
lib/node/actions/core.dart:17:17:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external void warning(String name);
                ^
lib/node/actions/core.dart:18:17:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external void error(String name);
                ^
lib/node/actions/core.dart:20:17:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external void addPath(String element);
                ^
lib/node/actions/core.dart:22:17:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external void exportVariable(String name, String value);
                ^
lib/node/actions/core.dart:24:17:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external void setFailed(String name);
                ^
lib/node/actions/core.dart:27:22:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external JSPromise getIDToken(String audience);
                     ^
lib/node/actions/exec.dart:11:14:
Error: Inline class 'Exec' is marked with a '@JS' annotation, but its representation type is not a valid JS interop type: 'InvalidType(<invalid>)'.
inline class Exec {
             ^
lib/node/actions/exec.dart:16:22:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external JSPromise exec(String commandLine, [List<String> args]);
                     ^
lib/node/actions/http_client.dart:8:14:
Error: Inline class 'HttpClient' is marked with a '@JS' annotation, but its representation type is not a valid JS interop type: 'InvalidType(<invalid>)'.
inline class HttpClient {
             ^
lib/node/actions/http_client.dart:9:12:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external HttpClient(
           ^
lib/node/actions/http_client.dart:16:22:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external JSPromise getJson(String requestUrl);
                     ^
lib/node/actions/tool_cache.dart:11:14:
Error: Inline class 'ToolCache' is marked with a '@JS' annotation, but its representation type is not a valid JS interop type: 'InvalidType(<invalid>)'.
inline class ToolCache {
             ^
lib/node/actions/tool_cache.dart:17:19:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external String find(
                  ^
lib/node/actions/tool_cache.dart:28:22:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external JSPromise downloadTool(String url);
                     ^
lib/node/actions/tool_cache.dart:34:22:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external JSPromise extractZip(String file);
                     ^
lib/node/actions/tool_cache.dart:43:22:
Error: Inline class member is marked 'external', but the representation type of its inline class is not a valid JS interop type.
  external JSPromise cacheDir(
                     ^
Error: Compilation failed.
```

**Comment by srujzs** [comment]
Nice, looking good! You're getting those errors because you have to declare a representation field/type on every inline class (now possibly called extension type!). For your purposes, it'll likely always be `JSObject`. Once the CFE makes it an error to not have a representation field, this will be more obvious.

File: lib/node/actions/http_client.dart [hunk 1]
```diff
@@ -6,12 +6,15 @@ import 'dart:js_interop';
 
 @JS()
 inline class HttpClient {
+  // TODO: This needs to be assigned somewhere.
```
Comment (devoncarew) [inline lib/node/actions/http_client.dart:9]: Adding this representation class field addresses the compilation issue, and the tests for this repo pass (all the `setup-dart` scenarios work), but it's a bit disconcerting that nothing assigns to this field.

Comment (srujzs) [reply lib/node/actions/http_client.dart:9]: The ergonomics of inline classes can be a bit weird. This field *is* initialized, but by the `external` constructor implicitly. If extension types + the primary constructor syntax is finalized, this becomes a little less confusing.

**Review by srujzs — APPROVED** [review]

File: lib/main.dart [hunk 2]
```diff
@@ -116,7 +116,10 @@ void main(List<String> args) async {
     }
 
     // Report success; print version.
-    await promiseToFuture(exec.exec('dart', ['--version']));
+    await promiseToFuture(exec.exec(
+      'dart',
+      ['--version'].map((str) => str.toJS).toList().toJS,
```
Comment (srujzs) [inline lib/main.dart:121]: Oh, forgot to mention - I think you can probably just do `['--version'.toJS].toJS`.

File: lib/node/actions/http_client.dart (hunk 1 above)
Comment (devoncarew) [reply lib/node/actions/http_client.dart]: Gotcha - sounds like not having explicit initialization isn't an issue.

File: lib/main.dart (hunk 2 above)
Comment (devoncarew) [reply lib/main.dart]: 👍  - simplified

**Comment by devoncarew** [comment]
cc @mit-mit for FYI re: the updates for the latest inline classes impl.

**Comment by mit-mit** [comment]
Thanks @devoncarew, looks nice

**Pull request merged**